In [1]:
import sqlite3
import pandas as pd

conn = sqlite3.connect("olist.db")
cursor = conn.cursor()

In [2]:
customers = pd.read_csv("olist_customers_dataset.csv")
geolocation = pd.read_csv("olist_geolocation_dataset.csv")
order_items = pd.read_csv("olist_order_items_dataset.csv")
order_payments = pd.read_csv("olist_order_payments_dataset.csv")
order_reviews = pd.read_csv("olist_order_reviews_dataset.csv")
orders = pd.read_csv("olist_orders_dataset.csv")
products = pd.read_csv("olist_products_dataset.csv")
sellers = pd.read_csv("olist_sellers_dataset.csv")
translation = pd.read_csv("product_category_name_translation.csv")

In [3]:
customers.to_sql("customers", conn, if_exists="replace", index=False)
geolocation.to_sql("geolocation", conn, if_exists="replace", index=False)
order_items.to_sql("order_items", conn, if_exists="replace", index=False)
order_payments.to_sql("order_payments", conn, if_exists="replace", index=False)
order_reviews.to_sql("order_reviews", conn, if_exists="replace", index=False)
orders.to_sql("orders", conn, if_exists="replace", index=False)
products.to_sql("products", conn, if_exists="replace", index=False)
sellers.to_sql("sellers", conn, if_exists="replace", index=False)
translation.to_sql("translation", conn, if_exists="replace", index=False)

71

In [4]:
print(customers.info(), geolocation.info(), order_items.info(), order_payments.info(), order_reviews.info(),
     orders.info(), products.info(), sellers.info(), translation.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  

In [5]:
customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [6]:
geolocation.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

In [7]:
order_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [8]:
order_payments.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [9]:
order_reviews.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

In [10]:
orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [11]:
products.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [12]:
sellers.head()

seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

In [13]:
translation.head()

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

## Main Question: What determines low vs high customer ratings in Olist e-commerce?

#### Key Explorations:
1. Sellers
2. Geolocation

#### Supporting variable
- Delivery time

In [14]:
sellers_query = pd.read_sql("""
WITH CTE AS (
    SELECT
        c.customer_id,
        s.seller_id,
        o.order_status,
        o.order_purchase_timestamp,
        o.order_approved_at,
        o.order_delivered_customer_date,
        o.order_estimated_delivery_date,
        ore.review_score
    FROM customers c
    JOIN orders o
        ON o.customer_id = c.customer_id
    JOIN order_items oi
        ON oi.order_id = o.order_id
    JOIN sellers s
        ON s.seller_id = oi.seller_id
    JOIN order_reviews ore
        ON ore.order_id = oi.order_id
)
SELECT
    *,
    (julianday(order_approved_at) - julianday(order_purchase_timestamp)) * 24
        AS approval_time_hours,
    julianday(order_delivered_customer_date) - julianday(order_estimated_delivery_date) AS delivery_delay_days
FROM CTE
""", conn)

In [15]:
sellers_query.to_sql("sellers_query", conn, if_exists="replace", index=False)

112372

In [16]:
sellers_final = pd.read_sql("""
SELECT
    seller_id,
    COUNT(*) as total_orders,
    AVG(approval_time_hours) AS AVG_approval_time_hours,
    AVG(delivery_delay_days) AS AVG_delivery_delay_days,
    AVG(review_score) AS AVG_review_score
FROM sellers_query
WHERE order_status = 'delivered'
GROUP BY seller_id
HAVING total_orders > 5
ORDER BY total_orders DESC
""", conn)

sellers_final.to_csv("sellers_final.csv", index=False)

In [17]:
sellers_final

seller_id  total_orders  AVG_approval_time_hours  \
0     6560211a19b47992c3666cc44a7e94c0          1985                 9.754353   
1     4a3ca9315b744ce9f8e9374361493884          1949                 9.819256   
2     1f50f920176fa81dab994f9023523100          1927                13.307414   
3     cc419e0650a3c5ba77189a1882b7556a          1755                11.422874   
4     da8622b14eb17ae2831f4ac5b9dab84a          1565                11.393742   
...                                ...           ...                      ...   
1715  0bfd0b434243adc5c2e1aa6baf19c6aa             6                 2.309769   
1716  0873d9f8f36123f8d910f4760e788cfb             6                 4.172824   
1717  077c5fae4bea9500e3737b16f71b9d3a             6                 9.440833   
1718  05feb94f19d094d4b0f9281f0b1d4c99             6                 4.118241   
1719  0307f7565ff85b299b6e5ef162b1b1a9             6                 2.212778   

      AVG_delivery_delay_days  AVG_review_score  
0                  -11.040620          3.951637  
1                   -8.962581          3.828630  
2                  -10.198338          3.987545  
3                  -12.327548          4.145869  
4                  -10.654535          4.075399  
...                       ...               ...  
1715               -19.809383          3.666667  
1716                -7.166553          3.166667  
1717                -9.081753          4.666667  
1718               -12.475008          4.500000  
1719               -16.086399          3.500000  

[1720 rows x 5 columns]

In [18]:
geolocation_query = pd.read_sql("""
WITH CTE AS (
    SELECT
        c.customer_state,
        o.order_status,
        o.order_purchase_timestamp,
        o.order_approved_at,
        o.order_delivered_customer_date,
        o.order_estimated_delivery_date,
        ore.review_score
    FROM customers c
    JOIN orders o
        ON o.customer_id = c.customer_id
    JOIN order_reviews ore
        ON ore.order_id = o.order_id
)
SELECT
    *,
    (julianday(order_approved_at) - julianday(order_purchase_timestamp)) * 24
        AS approval_time_hours,
    julianday(order_delivered_customer_date) - julianday(order_estimated_delivery_date)
        AS delivery_delay_days
FROM CTE
""", conn)

In [19]:
geolocation_query.to_sql("geolocation_query", conn, if_exists="replace", index=False)

99224

In [20]:
geolocation_final = pd.read_sql("""
SELECT
    COUNT(*) AS total_orders,
    customer_state,
    AVG(approval_time_hours) AS AVG_approval_time_hours,
    AVG(delivery_delay_days) AS AVG_delivery_delay_days,
    AVG(review_score) AS AVG_review_score
FROM geolocation_query
WHERE order_status = 'delivered'
GROUP BY customer_state
""", conn)

geolocation_final.to_csv("geolocation_final.csv", index=False)

In [21]:
geolocation_final

total_orders customer_state  AVG_approval_time_hours  \
0             80             AC                15.169441   
1            398             AL                12.034271   
2            145             AM                 9.761586   
3             66             AP                15.143114   
4           3246             BA                11.762253   
5           1276             CE                11.271846   
6           2089             DF                10.280527   
7           1978             ES                10.556748   
8           1963             GO                12.558030   
9            716             MA                13.851089   
10         11354             MG                 9.993606   
11           710             MS                10.510102   
12           882             MT                13.463118   
13           939             PA                12.221958   
14           513             PB                13.895830   
15          1590             PE                10.788619   
16           472             PI                10.161048   
17          4919             PR                10.820600   
18         12281             RJ                 9.299345   
19           473             RN                10.954472   
20           242             RO                13.380383   
21            41             RR                 9.492222   
22          5364             RS                11.672885   
23          3533             SC                11.124087   
24           334             SE                10.126417   
25         40484             SP                 9.735275   
26           273             TO                13.072419   

    AVG_delivery_delay_days  AVG_review_score  
0                -20.077007          4.087500  
1                 -8.147745          3.841709  
2                -18.866970          4.220690  
3                -19.091452          4.242424  
4                -10.190926          3.929452  
5                -10.173555          3.938088  
6                -11.302744          4.131642  
7                 -9.986137          4.082406  
8                -11.559897          4.102904  
9                 -9.036513          3.839385  
10               -12.576868          4.192707  
11               -10.320320          4.169014  
12               -13.805973          4.147392  
13               -13.497326          3.912673  
14               -12.801604          4.077973  
15               -12.735777          4.085535  
16               -10.607972          3.995763  
17               -12.638836          4.238260  
18               -11.104759          3.963439  
19               -13.005872          4.150106  
20               -19.385022          4.165289  
21               -16.594598          3.902439  
22               -13.215220          4.185496  
23               -10.850534          4.131050  
24                -9.376204          3.907186  
25               -10.406551          4.246196  
26               -11.414359          4.153846

In [22]:
sellers_final.to_sql("sellers_final", conn, if_exists="replace", index=False)

1720

In [23]:
pd.read_sql("""
SELECT
    AVG(AVG_review_score) AS avg_review_all_sellers
FROM sellers_final
""", conn)

avg_review_all_sellers
0                4.136487

In [24]:
pd.read_sql("""
WITH CTE AS (
    SELECT
        *,
        AVG(AVG_approval_time_hours) OVER () AS avg_review_approval_sellers,
        AVG(AVG_delivery_delay_days) OVER () AS avg_review_delivery_sellers
    FROM sellers_final
)
SELECT
    AVG(AVG_review_score) AS avg_review_fast_approval_sellers
    FROM cte
    WHERE AVG_approval_time_hours < avg_review_approval_sellers
""", conn)

avg_review_fast_approval_sellers
0                          4.163004

In [25]:
pd.read_sql("""
WITH CTE AS (
    SELECT
        *,
        AVG(AVG_approval_time_hours) OVER () AS avg_review_approval_sellers,
        AVG(AVG_delivery_delay_days) OVER () AS avg_review_delivery_sellers
    FROM sellers_final
)
SELECT
    AVG(AVG_review_score) AS avg_review_fast_delivery_sellers
    FROM cte
    WHERE AVG_delivery_delay_days < avg_review_delivery_sellers
""", conn)

avg_review_fast_delivery_sellers
0                          4.208723

In [26]:
pd.read_sql("""
WITH CTE AS (
    SELECT
        *,
        AVG(AVG_approval_time_hours) OVER () AS avg_review_fast_approval_sellers,
        AVG(AVG_delivery_delay_days) OVER () AS avg_review_fast_delivery_sellers
    FROM sellers_final
)
SELECT
    AVG(AVG_review_score) AS avg_review_fast_approval_and_delivery
    FROM cte
    WHERE AVG_delivery_delay_days < avg_review_fast_delivery_sellers AND AVG_approval_time_hours < avg_review_fast_approval_sellers
""", conn)

avg_review_fast_approval_and_delivery
0                                4.21188

The average review score across all sellers is approximately 4.14.

Sellers with faster-than-average approval times have a slightly higher average review score (≈ +0.02).

Sellers with faster-than-average delivery times show a more noticeable increase in review scores (≈ +0.06).

When sellers are faster than average in both approval and delivery, the average review score increases the most (≈ +0.07).

Conclusion:
While review scores are subjective and influenced by factors outside a seller’s control, faster delivery times appear to have a stronger relationship with higher customer reviews than approval speed. This is likely because approval times vary by hours, whereas delivery delays vary by multiple days, making delivery more salient to customers.